### PR TITLE
Fix of memory explosion, added blacklisting of libraries, a recursion default and fixed version number

### PR DESCRIPTION
**Pull request by noseglasses**
This PR that comprises multiple changes.

# Excessive memory consumption
When I ran this on the [Kaleidoscope project](https://github.com/keyboardio/Kaleidoscope), I was confronted with an excessive memory consumption (more than 32 Gb) due to a missing removal of duplicate include entries during recursive include processing. With a slight modification (removal of duplicates), memory consumption and configure times could be brought back to normal.

This considerably speeds up the configuration procedure.

# Blacklisting of libraries
Arduino-CMake looks for Arduino libraries whose names match those of include files encountered. Under some circumstances this leads to undesired behavior. An example is an include file `Keyboard.h` in a project that has no relation to Arduino's standard Keyboard library. To prevent the starndard library from being falsely added to the list of libraries to build, I added a cache variable that can hold a list of absolute paths of blacklisted libraries (for the example it would be `${ARDUINO_SDK_PATH}/libraries/Keyboard/src`).

The original behavior is not affected by this change.

# Recusion default
An additional cache variable `ARDUINO_CMAKE_RECURSION_DEFAULT` was introduced that enables to replace the default value `False` assigned for undefined `*_RECURSE` variables.

The original behavior is not affected by this change.

# Six digit Arduino version number 
ARDUINO_VERSION_DEFINE, originally a four digit number, was appended by the revision, now being a six digit number that allows for correct version checks.

This might affect others. If you want, I can introduce a configuration flag to preserve the original behavior.

### CODE REVIEW

**Comment by JonasProgrammer** [comment]
First of all, thanks for contributing to ArduinoCMake!

Your changes look good so far, but as per #10 the project is currently being refactored, thus making changes to the master is appreciated, but not ideal. Could you try to upstream-port your changes to feature/refactoring and PR into there?

**Comment by noseglasses** [comment]
Sorry, I currently lack time for such a complex operation. Unfortunately, at least in this case git is unable to properly track the migration of code to new files.

If I had been aware of the ongoing refactoring, I might have waited with my changes until the rebase is completed. 

**It might be useful for other users/devs to be informed about the refactoring at the top of the main README-page. Else you might receive a lot of PRs that are completely useless or hard to rebase.**

**Comment by fraillt** [comment]
For this reason, i strongly suggest to add CONTRIBUTING.md file to project, it will make life easier for contributors and maintainers.

**Comment by MrPointer** [comment]
@noseglasses Thank you for your time investment and of course for the great contribution that came out of it!
As I took the refactoring process to be my task (at least for now), I'll also be the one scratching my head about how your changes would be easily integrated. I'll update the process, and ask questions if required to, but mostly I just want to let you know that it is our mistake not PRing the refactoring earlier, or having a **Contributors** file as suggested by @fraillt.
Yet again - Thank you for your great contribution!

**Comment by JonasProgrammer** [comment]
@noseglasses Sorry for all the inconvenience, I'm carrying the burden of guilt here as well. I'm really not used to being a "maintainer" for OSS projects, the solely reason that I got one was that people suddenly started using my fork -- the original purpose of which was plain and simply to have something that would build an arduino firmware on a build server with a somewhat recent SDK.
The reason why this was yet again forked into an org was that MrPointer (who's the driving force) made huge contributions but I couldn't give him contributor rights due to some policy hiccups about write access to certain branches. Again, that doesn't apologize not having proper contribution guidelines etc., I just wanted to elaborate how I suddenly was assigned maintainer status of something that was intended to have 'works for me' status within roughly 2 weeks.

As suggested by by @fraillt in #20 we'll probably use semantic versioning from now onwards, so the current status can be retained as a 'v1' branch without any breaking changes.

@MrPointer Would you agree on merging his PR into the master which would turn into a v1-release branch by the time your refactored version is ready to become the new master? And for the basic problem of library recursion us fixing it independently in the upcoming v2?

**Comment by MrPointer** [comment]
@JonasProgrammer Yes, feel free to merge it to the master so other's will benefit from it while we complete the refactored version. As for the problems, I think I will be able to address them myself in the refactored version, so no need to duplicate code and effort.

**Review by JonasProgrammer — APPROVED** [review]
LGTM

File: cmake/Platform/Arduino.cmake [hunk 1]
```diff
@@ -2269,6 +2307,7 @@ function(get_arduino_flags COMPILE_FLAGS_VAR LINK_FLAGS_VAR BOARD_ID MANUAL)
             else ()
                 set(ARDUINO_VERSION_DEFINE "${ARDUINO_VERSION_DEFINE}0${CMAKE_MATCH_2}")
             endif ()
+            set(ARDUINO_VERSION_DEFINE "${ARDUINO_VERSION_DEFINE}00")
```
Comment (MrPointer) [inline cmake/Platform/Arduino.cmake:2310]: I don't quite understand what does it change, can you explain it some more? @noseglasses

**Comment by noseglasses** [comment]
The arduino version was only represented by its major and minor. Other tools require a six digit integer representation that also comprises the revision. My change is a simple fix for that. However, it would be even better to add two digits for the actual revision instead of `00`.
In any case I am not sure if this breaks other peoples' legacy code. It might be wise to wrap it in an if-clause that queries a global, e.g. `if(HAVE_SIX_DIGIT_VERSION)`.

**Comment by MrPointer** [comment]
Thanks, I'll try to figure out what's the best way to approach this, with your suggestion in mind.
BTW, can you please comment on the **Review Block** next time? It helps to keep things tighter.

**Pull request merged**